desc/singularities.py

```diff
@@ -0,0 +1,768 @@
+
+
+def _get_quadrature_nodes(q):
+    """Polar nodes for quadrature around singular point.
+
+    Parameters
+    ----------
+    q : int
+        Order of quadrature in radial and azimuthal directions.
+
+    Returns
+    -------
+    r, w : ndarray
+        Radial and azimuthal coordinates
+    dr, dw : ndarray
+        Radial and azimuthal spacing/quadrature weights
+
+    """
+    Nr = Nw = q
+    r, dr = scipy.special.roots_legendre(Nr)
+    # integrate separately over [-1,0] and [0,1]
+    r1 = 1 / 2 * r - 1 / 2
+    r2 = 1 / 2 * r + 1 / 2
+    r = jnp.concatenate([r1, r2])
+    dr = jnp.concatenate([dr, dr]) / 2
+    w = np.linspace(0, np.pi, Nw, endpoint=False)
+    dw = jnp.ones_like(w) * np.pi / Nw
+    r, w = jnp.meshgrid(r, w)
+    r = r.flatten()
+    w = w.flatten()
+    dr, dw = jnp.meshgrid(dr, dw)
+    dr = dr.flatten()
+    dw = dw.flatten()
+    return r, w, dr, dw
+
+
+class _BIESTInterpolator(IOAble, ABC):
+    """Base class for interpolators from cartesian to polar domain.
+
+    Used for singular integral calculations.
+
+    Parameters
+    ----------
+    eval_grid, source_grid : Grid
+        Evaluation and source points for the integral transform.
+        source_grid should be a LinearGrid
+    s : int
+        Extent of polar grid in number of src grid points. Same as "M" in the
+        original Malhotra papers.
+    q : int
+        Order of quadrature in polar domain
+
+    """
+
+    _io_attrs_ = ["_eval_grid", "_source_grid", "_q", "_s"]
+
+    @abstractmethod
+    def __init__(self, eval_grid, source_grid, s, q):
+        pass
+
+    @property
+    def s(self):
+        """int: Extent of polar grid in number of src grid points."""
+        return self._s
+
+    @property
+    def q(self):
+        """int: Order of quadrature in polar domain."""
+        return self._q
+
+    @abstractmethod
+    def __call__(self, f, i):
+        """Interpolate data to polar grid points.
+
+        Parameters
+        ----------
+        f : ndarray
+            Data at source grid points to interpolate.
+        i : int
+            Index of polar node.
+
+        Returns
+        -------
+        fi : ndarray
+            Source data interpolated to ith polar node.
+        """
+
+
+class FFTInterpolator(_BIESTInterpolator):
+    """FFT interpolation operator required for high order singular integration.
+
+    Parameters
+    ----------
+    eval_grid, source_grid : Grid
+        Evaluation and source points for the integral transform.
+        Both should be LinearGrid without stellarator symmetry.
+    s : int
+        Extent of polar grid in number of src grid points. Same as "M" in the
+        original Malhotra papers.
+    q : int
+        Order of quadrature in polar domain
+
+    """
+
+    _io_attrs_ = _BIESTInterpolator._io_attrs_ + ["_h_t", "_h_z", "_st", "_sz"]
+
+    def __init__(self, eval_grid, source_grid, s, q):
+        # current fft interpolating can't handle symmetric grids correctly
```

Comment:
Is this something we could fix in the future? If so, make sure we have an Issue for it so we don't forget. I assume it could be a substantial speedup.